[2.4] Add support for missing Curl methods to the Curl client (#40157)

lbajsarowicz opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
Cherry-picked change from #39330 rated as "backwards-incompatible" by @engcom-Dash rejecting all the changes to Testing Framework, leaving just the new methods in the `Curl` class.

_It makes me upset that after spending hours to get the previous PR in shape to merge, this **bug fix** was taken down instead of recommendation to fix imaginary "backwards incompatiblity", provided Maintainers can point what change became backwards-incompatible._

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. ...
2. ...

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)


### Resolved issues:
1. [x] resolves magento/magento2#40157: [2.4] Add support for missing Curl methods to the Curl client

lbajsarowicz committed:
SwiftOtter-SOP-348 Match CURL Client to RFC-9110 and RFC-5789

lbajsarowicz commented:
@magento run all tests

lbajsarowicz committed:
SwiftOtter-SOP-348 Pass HTTP payload `curlopt`

lbajsarowicz commented:
@magento run all tests (after changing the method to pass payloads)

bgorski commented:
@magento run all tests

lbajsarowicz committed:
SwiftOtter-SOP-348 Simplify method signature to match TestFramework's ("backwards compatibility")

lbajsarowicz commented:
@magento run all tests

I simplified the `delete` method signature to make the method compatible to the `Magento\TestFramework`, which - according to @engcom-Dash (in #39330) has higher priority than `Magento\Framework`.

bgorski commented on lib/internal/Magento/Framework/HTTP/Client/Curl.php:
```diff
@@ -239,13 +250,109 @@ public function get($uri)
+     */
+    public function delete($uri)
+    {
+        $this->makeRequest("DELETE", $uri, $params);
```
@lbajsarowicz there's a bug in here - it uses $params for makeRequest but is not receiving any in the method head, there's only $uri there.

lbajsarowicz replied:
Nice catch. I'm handling it.

lbajsarowicz committed:
SOP-348 #36336 Fix invalid `DELETE` method reference (no payload)

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented on lib/internal/Magento/Framework/HTTP/Client/Curl.php:
```diff
@@ -239,13 +249,108 @@ public function get($uri)
+     *
+     * @url https://www.rfc-editor.org/rfc/rfc9110.html#section-9.3.5
+     */
+    public function delete($uri)
```
Add `void` as the function return type 
```suggestion
    public function delete($uri): void
```

engcom-Hotel replied:
As per the [documentation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/DELETE), delete method might have Payload. So I think we can add `$params` as an optional variable. I think here @bgorski is just pointing out that we are not passing `$params` but using it in the makeRequest method.

Den4ik replied:
@engcom-Hotel `$params` will never be used in `makeRequest` for `DELETE` method. I think that it is not worth adding this parameter to the function description, because it will only confuse

engcom-Hotel replied:
Hello @Den4ik, actually we are referring to this [documentation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/DELETE) for `DELETE` method. And according to this, `DELETE` method might have Request. Thats I made this review comment.

Den4ik replied:
Hi @engcom-Hotel,

Thanks for your comment. Please compare the documentation for the DELETE request with the [documentation for the GET request](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/GET):

> An optional query component preceded by a question mark (?)

If you look at the GET request implementation, you will see that it doesn't contain any parameters.
The DELETE request implementation follows the same style as the GET request, and query parameters (preceded by a question mark) should be included in $uri.

engcom-Hotel reviewed (changes requested):
Hello @lbajsarowicz,

Thanks for the contribution!

Please have a look at the below review comments.

Thanks

engcom-Charlie commented:
Hi @lbajsarowicz,

Did you get chance to look into [these](https://github.com/magento/magento2/pull/39471#pullrequestreview-2558523669) review comments.

lbajsarowicz commented:
**Update** Looking into it.

Den4ik reviewed (approved):
Hi @lbajsarowicz,
Thanks for your work.
In my opinion this is really good candidate for 2.4.8 release.

engcom-Charlie committed:
Worked on review comments

engcom-Charlie commented:
Hi @lbajsarowicz,

As mentioned [here](https://github.com/magento/magento2/pull/39471#issuecomment-2633371359) can you please work on the review comments. I have worked on it but unable to push the changes because of access rights issue to your repository. 

Please do the needful so that we can able to proceed further on this PR.

lbajsarowicz commented:
@engcom-Charlie I'm working on it now.

lbajsarowicz committed:
Code Review / Adjust `delete` method declaration

lbajsarowicz commented:
@magento run all tests

lbajsarowicz committed:
Code Review / Adjust `delete` method declaration

bgorski reviewed (approved):
@lbajsarowicz thank you for the contribution! 
@engcom-Charlie with the $params issue handled, this is ready to get processed further.

lbajsarowicz commented:
@engcom-Charlie  I've discussed the `$params` with @bgorski offline and we have agreed that it's easier to change the implementation to support `$params` in the future, if there's such a case, rather than modifying already published API for `delete` method (ergo: Look how many issues the Integration Tests framework caused).

engcom-Charlie commented:
As we have got the review approval from @bgorski moving this PR to Ready for testing.

engcom-Charlie commented:
@magento run all tests

Den4ik commented:
Returning to the question of using `$params`, this parameter does not serve any specific purpose and is misleading.  

I have an image located at `domain.test/media/catalog/product/some_product.jpg`. I use Lambda and S3 for image resizing, for example:  
`domain.test/media/catalog/product/some_product.jpg?width=255&height=255`.  

Let's say I want to delete an image of a specific size but keep the original. I see that the `delete` method accepts parameters, so I call:  
```php
delete('domain.test/media/catalog/product/some_product.jpg', ['width' => 255, 'height' => 255]);
```
As a result, I end up with something completely different from what I expected.

lbajsarowicz commented:
I understand your idea - but still the standard mentions the existence of params. Imagine that one day Google introduces this feature of the standard. 

Change to Magento would be more complicated than introducing that for 8 billion Internet users around the world, as this would require method signature changes.

This is also why we (with @bgorski) agreed after discussion that the 2nd argument should be default to [] to avoid unnecessary confusion.

engcom-Charlie commented:
Manual testing is not required here, as this PR is taking care of few missed CURL methods to existing CURL client. Since the builds are failing, moving it to Extended Testing.

engcom-Charlie commented:
@magento run Database Compare,  Functional Tests B2B, Functional Tests CE, Functional Tests EE, Semantic Version Checker, Static Tests, WebAPI Tests

engcom-Charlie commented:
@lbajsarowicz can you please have a look into [Static](https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/39471/6f4feec56becfafc3f99d09b062a1c78/Statics/console-error-logs.html) test failure.

For SVC failure, I have raised internal approval [JIRA](https://jira.corp.adobe.com/browse/AC-14282). Moving this PR to Pending Approval now. We will proceed ahead on this once will get all the required approvals.

engcom-Charlie commented:
> @engcom-Charlie It requires date in Copyrights.
> 
> What year you like? I like 1989 because that's the year when we got rid of communism in Poland.
> 
> Honestly, I don't know why its on Contributors to fix the Copyrights issues?

Hi @lbajsarowicz,

You can update the copyright with below format.

```
Copyright [year code created] Adobe 
All Rights Reserved.
```

lbajsarowicz commented:
@magento run all tests after this critical change.

engcom-Charlie commented:
@magento create issue

lbajsarowicz committed:
Add DELETE to HTTP_METHODS_WITH_PAYLOAD for body support

Enables delete() method to send optional $params as request body payload.
RFC 9110 allows payload on DELETE (undefined semantics, not prohibited).

lbajsarowicz commented:
@engcom-Charlie I've added the `$params` for `delete()` although I still don't think it's in-line with the RFC.

lbajsarowicz committed:
Merge branch '2.4-develop' into 2.4-develop-missing-curl-methods

lbajsarowicz commented:
@engcom-Charlie Any updates here?

engcom-Charlie commented:
> @lbajsarowicz we still waiting for all the required approvals, will update here once we get any further updates.

@lbajsarowicz will update here once we get any further updates. Thank you!
